python/private/python.bzl

```
@@ -104,7 +153,14 @@ def parse_modules(*, module_ctx, _fail = fail):
                 # * rules_python needs to set a soft default in case the root module doesn't,
                 #   e.g. if the root module doesn't use Python itself.
                 # * The root module is allowed to override the rules_python default.
-                is_default = toolchain_attr.is_default
+                if default_python_version:
+                    is_default = default_python_version == toolchain_version
+                    if toolchain_attr.is_default and not is_default:
+                        fail("The 'is_default' attribute doesn't work if you set " +
```

```suggestion
                        logger.warn("The 'is_default' attribute doesn't work if you set " +
```

Maybe having the logger warn the user is enough here? I imagine people may have a `.python-version` file that they expect bazel to not read and then suddenly `rules_python` will start failing, which would be surprising behaviour.

---

If we enter this branch, wouldn't there be surprising behavior no matter what? If the user has said `python.toolchain(python_version = "3.11", is_default = True)` and the `.python-version` file says `3.12`, suddenly `rules_python` will start using `3.12` even though the user has explicitly asked for `3.11` to be the default. It's not that some obscure logic used to choose 3.11 and now some obscure logic chooses 3.12 and that's worth a warning — the user explicitly asked for something that we're now actively refusing to do. That feels more like a `fail` than a `logger.warn`, doesn't it? In this scenario, if the user wants `rules_python` to ignore the `.python-version` file and keep using `3.11`, they would have to migrate to the `python.defaults(python_version = "3.11")` syntax and stop using `is_default`. I suppose this may be considered a breaking change, which we might need to document better in the change log and perhaps the failure message or warning. If we don't want a breaking change, we'd have to change the logic so the `is_default` overrides the `.python-version` file: in that case, I think a warning would be suitable.

If we enter this branch because the user is using both `python.defaults(python_version = "3.12")`  and `python.toolchain(python_version = "3.11", is_default = True)`, I think that should be a `fail` rather than a `logger.warn`, because the user is explicitly asking for things that aren't consistent. We could try to keep track of whether the `.python-version` file or the `python.defaults` tag class is the source of the discrepancy and do different error messages (or warning, perhaps, in the case of the `.python-version` file).

---

Thanks, makes sense.

---

We can skip the surprising behavior if we don't read `.python-version` unbidden, so I stop trying that in 5bdfbe2 (see commit message).